### ISSUE TITLE
`modus-themes` to succeed `modus-operandi-theme` and `modus-vivendi-theme`

### ISSUE THREAD

**Issue by protesilaos** [melpa/melpa#7245]
## Plan for the transition

I wish to introduce a new `modus-themes` package as the successor to `modus-operandi-theme` and `modus-vivendi-theme` (I will now prepare a pull request that adds the new recipe and edits the existing ones).

My proposal for the transition is this:

1. Add `modus-themes`.  It will be receiving all development attention going forward.

2. Stop updating `modus-operandi-theme`, `modus-vivendi-theme`, but leave them in place for a reasonable amount of time (maybe 6 months to a year?), so that users are given enough time to manually change their sources.  Delete them afterwards.

The code for the `modus-themes` is here: <https://gitlab.com/protesilaos/modus-themes/-/tree/main>.

## Historical context

About a year ago I added the packages `modus-operandi-theme` and `modus-vivendi-theme`.  These are part of the same project but were distributed as standalone items.  The reason at the time was that they were not derived from the same code base, while I did not have the requisite Elisp knowledge to implement such a design.

Fast-forward to the present, where I have made all the necessary preparations for distributing the themes as a single package.  They now derive from a common base.

## Other considerations

1. The download counts on melpa.org might indicate that users prefer to have both themes available (data from 2020-11-23 at 13:17 +0300), even though I understand this is a problematic assumption if you want to be strict about it:

| Package              | Downloads |
|----------------------|-----------|
| modus-operandi-theme | 56,404    |
| modus-vivendi-theme  | 56,000    |

2. The refactored code makes it easier for me to continue developing the project, while it unifies customisation options and, more generally, cuts down on a lot of duplicate code on my side as well as that of users.

**Comment by tarsius** [comment]
> Stop updating `modus-operandi-theme`, `modus-vivendi-theme`, but leave them in place for a reasonable amount of time (maybe 6 months to a year?), so that users are given enough time to manually change their sources.  Delete them afterwards.

While that is a noble approach, in practice this just means that the issue won't pop up now but instead in 6 months to a year. As far as I know neither Melpa nor `package.el` can communicate to the user that some installed package was removed from the elpa archive whence it came. So we usually use the "rip of the bandage approach" in these parts.

**Comment by minad** [comment]
@protesilaos @tarsius I would very much prefer to have only one modus package. And if that helps @protesilaos in maintaining modus, this should be merged. Is it possible to overwrite modus-operandi-theme and modus-vivendi-theme with a small package which depends on modus for compatibility and prints some kind of deprecation warning, maybe via the autoload mechanism?

**Comment by tarsius** [comment]
> Just delete the old packages directly?

Yes.

**Comment by protesilaos** [comment]
>> Just delete the old packages directly?
>
> Yes.

Fine by me!  Do I just delete the recipes and update [my PR](https://github.com/melpa/melpa/pull/7246)?  Or is this something only maintainers can do?

**Comment by tarsius** [comment]
Just update the existing pr.

**Comment by tarsius** [comment]
Please also update GNU Elpa.

**Comment by protesilaos** [comment]
Thank you Jonas!  I will do it, though there is a blocking issue right now: the themes in core Emacs need to be updated to the latest version.  This requires some changes to how built-in themes work ([read this for a recap](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=45068#86)).  I will try to come up with a solution but it might take some more time.  Sorry about that!

The reason this is blocking the progress on GNU Elpa is that those themes must be updated to pull directly from Emacs instead of my repo.  So once the above is done, this will happen as well.

That granted, I have not forgotten [what Chris told me](https://github.com/melpa/melpa/pull/7246#issuecomment-745578490), namely, to remove the Melpa packages in order to avoid confusion.  I would just need some more time, though I understand if you want to speed things up on your end.

**Comment by tarsius** [comment]
Thanks for the information. I'm not in a hurry but still wanted to make sure this is still being pursued.

**Comment by protesilaos** [comment]
Hello again!  GNU ELPA has just been updated to include the new `modus-themes` package: https://elpa.gnu.org/packages/modus-themes.html. Though we still need to delete the old packages from there.  I plan to work on this in the coming days.

I will also prepare a pull request to remove the package from Melpa, as we have already discussed.  Expect this by the end of this week or the beginning of the next.

**Comment by protesilaos** [comment]
Hello again Jonas!  I ping Chris @riscy, as we discussed this in [issue #7246](https://github.com/melpa/melpa/pull/7246#issuecomment-739329755).

To recapitulate:

+ The `modus-themes` package is available on MELPA and GNU ELPA.

+ The themes are also built into Emacs28, though are still maintained in my own repo.  GNU ELPA is instructed to pull from emacs.git when an update happens there (upon the release of a new version of the themes I send a patch to the Emacs maintainers).

+ The old packages `modus-{operandi,vivendi}-theme` were deleted from MELPA, while they have been replaced with transient packages on GNU ELPA which pull in the new `modus-themes`.  We will delete them from there in a few months.

+ `M-x list-packages` gives us the following:

```
  modus-operandi-theme     0.13.1         available    gnu      Accessible light theme (WCAG AAA)
  modus-themes             1.2.4          available    gnu      Highly accessible themes (WCAG AAA)
  modus-themes             20210319.543   available    melpa    Highly accessible themes (WCAG AAA)
  modus-vivendi-theme      0.13.1         available    gnu      Accessible light theme (WCAG AAA)
```

+ It is this potential source of confusion that prompted the discussion of deleting `modus-themes` from MELPA.

+ In principle I am willing to comply with your request and, generally, not pose any problems to the maintenance of this archive.  So I have stated that I will create a PR to remove the package.  However, we will then be in a position where it would no longer be possible for users to get the latest version of my project using `package.el`.  The other sources only distribute the latest tagged release.  So I would be missing out on potential feedback during development cycles.

+ Thus, I must ask: is it possible to reconsider and keep `modus-themes` on MELPA?  If that is not possible, I will open the PR.

Thanks again for your time!

**Comment by riscy** [comment]
Let me also mention @purcell but I think this strategy isn't a good idea.  Let me rephrase so I'm sure I've got it right -- the idea is to use MELPA to host unstable releases, and use ELPA to host stable (tagged) releases?

I haven't worked with ELPA before so please forgive any ignorance, but here are the issues I see:

1. Most importantly, this makes your package idiosyncratic and so users require extra instructions ("if you have installed modus-themes but want the latest development version, make sure you subscribe to the MELPA package index and pull from there instead")
2. Developers will also need extra instructions (where do I go for the canonical version of the code?)
3. It seems like it's more work for you; you have time for modus-themes now, but if you get ambushed with other work is there a risk this system could lead the project into a confusing state? (e.g. if a new maintainer volunteered to take over the project, how would they resolve this?)
3. There is a subtle issue with version numbers that this could exacerbate; if a user has modus version 20190320.1234 from MELPA installed, and then removes MELPA (unstable) from their configuration, my understanding is that they will never upgrade unless you use a compatible tagging scheme on ELPA.  This is also a problem with MELPA unstable/stable, but at least the problem is contained under one roof.

That's all I can think of right now.

**Comment by minad** [comment]
@riscy I am one of those users who would like to have access to the rolling release version of modus-themes via MELPA, since this allows fast feedback cycles between me as a user/maintainer of other packages and @protesilaos. The modus-themes are  stable, even if you consider the bleeding edge version, so I don't expect downsides for the users. There are multiple other packages which are available on both ELPA and MELPA and which follow a similar policy - stable versions are on ELPA and MELPA can be used for the rolling release version, which reflects the latest commit. Why is this suddenly a problem?

**Comment by riscy** [comment]
> Why is this suddenly a problem?

This is an area I don't have much experience with so I'll concede I may be wrong (hence the mention above).  But it may depend on how old those other packages are and whether the authors even communicated what they were doing to us in the first place.

**Comment by protesilaos** [comment]
> But it may depend on how old those other packages are

While we wait for further feedback I will try to clarify those points:

1. The other packages, i.e. `modus-{operandi,vivendi}-theme` were 4+ months old.  Now they have been replaced by a simple snippet that just pulls in the new `modus-themes` package.  Those old packages will be removed within the next few months: the idea is to ease the transition (this was requested by the maintainers, as my original patch was simply deleting the old ones and adding the new one).

2. Otherwise all packages are up-to-date: emacs.git and elpa.git contain the last tagged release, while here we have the latest commit.

> and whether the authors even communicated what they were doing to us in the first place.

Perhaps it is best to outline the sequence of events:

+ In November 2019 I submitted here `modus-{operandi,vivendi}-theme` as two standalone packages.

+ A few months later (I think March 2020) some users asked me whether I would be willing to also put them on ELPA.  Seeing as I was the only contributor at the time, I thought it would be okay to distribute them from there.  So I signed the FSF papers and went through the process.

  - I was not aware of any potential problem between ELPA and MELPA.  If it was documented somewhere, I either failed to read it or understand it.  My apologies for that.

+ In August 2020 the themes were incorporated in emacs.git upon the request of Emacs maintainers.  For the time being, development continues on my repo and I sync with Emacs whenever I publish a new tagged release.

+ In December 2020 I refactored the code of my project to eventually derive the two themes from a single source.  This led to the original message on this thread, where we eventually decided to delete the two packages I had contributed to MELPA on November 2019 and replace them with a single `modus-themes` package.

I hope this clarifies any questions you may have.  If something is unclear, please let me know and I will elaborate on it.

**Comment by riscy** [comment]
I was referring to @minad's reference to other packages (besides modus) that were already on both.  But thanks, that's still useful context. :)

**Comment by minad** [comment]
@riscy I understand that there can be confusion and that there may be technical difficulties. I consider my point about the treatment of older packages a weak argument actually. It would be nice if MELPA is consistent with regards to all packages, but this is hard to implement since policies change, one doesn't want to disrupt users unnecessarily and so on.

But I would like to stress the point about the fast release cycle. This is an amazing feature if a package is distributed via MELPA. It can be abused easily and lead to stability problems, but if used carefully it is great. When adding features/fixes in my [Consult](https://melpa.org/#/consult) package, people often tell me that they wait until the commit hits MELPA before giving feedback. This really lowers the barrier for contributions. Obviously this does not mean that one should outsource testing to the users. In the case of the modus-theme I think this point is very strong, since modus essentially tries to cover *every package*, which means that @protesilaos gets [a lot of feedback on his issue tracker](https://gitlab.com/protesilaos/modus-themes/-/issues). So I think this greatly outweighs the costs of having the package distributed via two channels. One could also argue that the ELPA people should offer such a possibility on their side, but it seems to be more pragmatic to just allow MELPA/ELPA distribution.

**Comment by tarsius** [comment]
>> Why is this suddenly a problem?

> This is an area I don't have much experience with so I'll concede I may be wrong (hence the mention above). But it may depend on how old those other packages are and whether the authors even communicated what they were doing to us in the first place.

I suppose I have more experience with it because I maintain the Emacsmirror, which tries to stay in sync with Melpa, GNU Elpa, Emacs etc. so conflicts between these things are potentially a problem. In practice these problems can be quickly resolved, unless upstream is unresponsive. I don't expect @protesilaos to become unresponsive any time soon, so I have no objects to making an exception in this case.

By the way, if I remember correctly, there is work being done to publish snapshots on GNU Elpa as well. Once that is released, Melpa should drop all packages that are distributed in both registries.

Also IMO you should request commit access to Emacs, @protesilaos.

**Comment by purcell** [comment]
So if I understand this correctly, ELPA will distribute copies of the code which ultimately come from emacs.git, and they will get into that repo by being copied across. In the meantime, ongoing development happens in the author's own github repo. The desire is to have stable copies come from ELPA, and snapshots from MELPA. Well, we can certainly accommodate that, and do so for other packages which exist in ELPA (e.g. mmm-mode). By default, we will also populate MELPA Stable with stable packages. If the source repo is tagged appropriately, everything should match up. In _theory_ the stable ELPA and MELPA Stable packages could diverge, if tagging/copying is not done cleanly. If *really* necessary we can suppress the MELPA Stable packages (by specifying a bogus tag pattern in the recipe), but I'd prefer not to if we can avoid it.

Historically most authors who distribute via ELPA don't want/need MELPA packages, but there's no hard and fast rule about that, and it's a matter of what the authors and users want.

**Comment by protesilaos** [comment]
> tarsius: I suppose I have more experience with it because I maintain the Emacsmirror, which tries to stay in sync with Melpa, GNU Elpa, Emacs etc. so conflicts between these things are potentially a problem. In practice these problems can be quickly resolved, unless upstream is unresponsive. I don't expect @protesilaos to become unresponsive any time soon, so I have no objects to making an exception in this case.

I see, thank you!  Indeed, there is no plan to abandon the project: I am continuously iterating on it and a new version will be released some time in mid-April.

> tarsius: By the way, if I remember correctly, there is work being done to publish snapshots on GNU Elpa as well. Once that is released, Melpa should drop all packages that are distributed in both registries.

There is an [elpa-devel variant](https://elpa.gnu.org/devel/index.html), which should produce development snapshots.  For me to use this I need to move development of the themes directly to emacs.git since ELPA pulls from there.  Though this is not my prerogative.

> tarsius: Also IMO you should request commit access to Emacs, @protesilaos.

I did gain access a few days ago (covers elpa.git and emacs.git), though I have only made one minor commit that appends "DEPRECATED" to the package description of the old packages.  When I publish the new version of the themes and prepare the corresponding patch for emacs.git, I will ask the maintainers what they think about moving development there.  I still have a lot to learn and am intimidated by the challenge and how fast things are moving, so I will not act unilaterally and only follow their recommendations.

If development of the themes moves to emacs.git and if we can get elpa-devel running from there, then I will contact you right away so that we can resolve potential conflicts with MELPA.  Otherwise we keep things as they are, until further notice.

> purcell: So if I understand this correctly, ELPA will distribute copies of the code which ultimately come from emacs.git, and they will get into that repo by being copied across. In the meantime, ongoing development happens in the author's own github repo. The desire is to have stable copies come from ELPA, and snapshots from MELPA. Well, we can certainly accommodate that, and do so for other packages which exist in ELPA (e.g.  mmm-mode). By default, we will also populate MELPA Stable with stable packages. If the source repo is tagged appropriately, everything should match up. In theory the stable ELPA and MELPA Stable packages could diverge, if tagging/copying is not done cleanly. If really necessary we can suppress the MELPA Stable packages (by specifying a bogus tag pattern in the recipe), but I'd prefer not to if we can avoid it.

Yes, that is the gist of it.  I will leave MELPA Stable up to you and will do whatever you think is best.

> purcell: Historically most authors who distribute via ELPA don't want/need MELPA packages, but there's no hard and fast rule about that, and it's a matter of what the authors and users want.

I want to share stable versions and snapshots of my project.  The added dimension of emacs.git has made things more complex, though this might be temporary.

I also wish to be a good citizen and cooperate with you.  If/when I have any update on the development of the project, I will promptly communicate with you.

Thanks again for your time and patience!

**Comment by tarsius** [comment]
> There is an elpa-devel variant, which should produce development snapshots.

Thanks for the link. I think I saw this before but cannot remember how I got there. Was this announced somewhere and/or are there any links to it? At least it's in my bookmarks now. :smile:

**Comment by protesilaos** [comment]
> I think I saw this before but cannot remember how I got there. Was this announced somewhere and/or are there any links to it?

It was discussed a while ago in emacs-devel as part of a series of internal changes to elpa.git.  But I do not think anything has been announced yet, so I am not sure what its official status is.
